chemreg/compound/serializers.py

```diff
@@ -64,21 +67,46 @@ class Meta:
             "smiles",
         )
         extra_kwargs = {"molfile_v3000": {"required": False, "trim_whitespace": False}}
-        validators = [
-            OneOfValidator("molfile_v2000", "molfile_v3000", "smiles", required=True)
-        ]
 
     def __init__(self, *args, admin_override=False, **kwargs):
         if not admin_override:
-            for structrue in ("molfile_v2000", "molfile_v3000", "smiles"):
-                field = self.fields[structrue]
+            for structure in self.alt_structures:
+                field = self.fields[structure]
                 field.validators.append(validate_inchikey_unique)
         super().__init__(*args, **kwargs)
 
-    def validate(self, data):
-        alt_structures = ("molfile_v2000", "molfile_v3000", "smiles")
-        alt_structure = next(k for k in alt_structures if k in data)
-        data["molfile_v3000"] = get_molfile_v3000(data.pop(alt_structure))
+    def validate_cid(self, value):
+        if "inchikey" not in self.initial_data:
+            raise ValidationError("InchIKey must be included when CID is defined.")
+        return value
+
+    def validate_inchikey(self, value):
+        if "cid" not in self.initial_data:
+            raise ValidationError("CID must be included when InchIKey is defined.")
+        return value
+
+    def to_internal_value(self, data):
+        matched_fields = set(self.alt_structures) & set(self.initial_data.keys())
+        formatted_fields = [format_value(f) for f in sorted(self.alt_structures)]
+        formatted_matched_fields = [format_value(f) for f in sorted(matched_fields)]
+        if not matched_fields:
+            raise ValidationError(
+                {"non_field_errors": f"One of {sorted(formatted_fields)} required."}
+            )
+        if len(matched_fields) > 1:
+            raise ValidationError(
+                {
+                    "non_field_errors": (
+                        f"Only one of {formatted_fields} allowed. "
+                        f"Recieved {formatted_matched_fields}."
+                    )
+                }
+            )
+        data = super().to_internal_value(data)  # calls field validators
+        structure = next(k for k in self.alt_structures if k in data)
+        data["molfile_v3000"] = get_molfile_v3000(data.pop(structure))
+        if "inchikey" not in data:
+            data["inchikey"] = get_inchikey(data["molfile_v3000"])
```

Comment:
Can this be moved to `validate`? I remember finding some wonky things happening when we did this kind of validation in `to_internal_value`. I think of `validate` like a `clean` and `to_internal_value` like a low-level transformation of data (e.g. a date-time string to a Python `DateTime` object.

Reply:
Because we have any one of three fields going into `molfile_v3000`, in the event that we don't have an `inchikey` we have to `get_inchikey` to pass to be serialized. So, in order to know that we are running `get_molfile_v3000` and `get_inchikey` on the right given structure there would have to be only one provided. Using the `OneOfValidator` will work here, but what seemed strange to me was if 2+ of the fields were provided and there was an error with a field, say "InchiKey already exists", that error could come up when providing 2 structures, but wouldn't identify the problem that there are 2 structures provided. And no matter what one will always be converted to v3000 because we need that to create the inchikey, so in the error message where v2000 and SMILES are provided it will say it received v3000 and smiles becuase we need to convert the structure before creating the inchikey.

Reply:
I made the initial decision to handle this step in `to_internal_value` because of what Rick has explained - we need to sort through the provided fields and check their results before running `validate`. There might be a different approach possible, but I didn't find a way to handle all the branching conditional logic in `validates`.